Dotnet Tool impementation

nkolev92 opened the pull request:
## Bug
Fixes: https://github.com/NuGet/Home/issues/6199, https://github.com/NuGet/Home/issues/6198, https://github.com/NuGet/Home/issues/6197, https://github.com/NuGet/Home/issues/6200

Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 
Here's the spec for the global tools implementation.
https://github.com/NuGet/Home/wiki/Global-Tools---NuGet-Implementation

The brief is, we introduce a new package type for dotnet tools, with compatible asset as "tools/tfm/rid/any.extension. 

That package is only compatible with a special project type that the CLI will create, DotnetToolReference. 
1 package per project, that package has to be a dotnettool.
DotnetTooReference works almost exactly the same way as Package Reference with the above restrictions in place.
Notable: Tool Restore does not generate Msbuild props and targets, for failures
No no-op artifacts for tool restore. 

The core logic is in:
CompatibilityChecker
CompatibilityIssue
RestoreCommand
LockFileUtils
LockFileFormat
LockFileTargetLibrary
ManagedCodeConventions

Most relevant tests:
DotnetToolTests
ContentModelToolsTests


## Testing/Validation
Tests Added: Yes 
Reason for not adding tests:  
Validation done:  Manual, by Will from CLI team, automation

PS. This will go in dev when it's shifted to 15.7
//cc @wli3 @rrelyea

nkolev92 committed:
add project style and package type

nkolev92 committed:
Add managed code conventions for tools

nkolev92 committed:
more wokr

nkolev92 committed:
add targets

nkolev92 committed:
msbuild fast

nkolev92 committed:
spec validation

nkolev92 committed:
cleanup nuget.targets to get execution

nkolev92 committed:
work

nkolev92 committed:
block dotnet tool reference projects with more than 1 reference

nkolev92 committed:
work

nkolev92 committed:
bla

nkolev92 committed:
fix

nkolev92 committed:
cleanup

nkolev92 committed:
fix test

nkolev92 committed:
fix test

nkolev92 committed:
cleanup

nkolev92 committed:
cleanup

nkolev92 committed:
cleanup

nkolev92 committed:
make the code compile :)

nkolev92 committed:
only block direct tools dependencies

nkolev92 committed:
Add code convention tests

nkolev92 committed:
work

nkolev92 committed:
fix is package compatible check

nkolev92 committed:
work

nkolev92 committed:
remove debugger

nkolev92 committed:
fix compat checker

nkolev92 committed:
add more tests, including 3 failing ones

nkolev92 committed:
platform theory

nkolev92 committed:
dsadas

nkolev92 committed:
work

nkolev92 committed:
collect all files in the tools group

nkolev92 committed:
fix tests

nkolev92 committed:
add lock file validation

nkolev92 committed:
add tests to assert json files are in the targets files

nkolev92 committed:
add runtime json test

nkolev92 committed:
add folder

nkolev92 committed:
add more tests, runtime any and similar etc.

nkolev92 committed:
more tests

nkolev92 committed:
fix pack tests broken due to CLI upgrade.

nkolev92 committed:
fix tests

nkolev92 commented on src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreJob.cs:
```diff
@@ -345,13 +345,6 @@ internal sealed class SolutionRestoreJob : ISolutionRestoreJob
             }
         }
 
-        private bool IsProjectBuildIntegrated(PackageSpec packageSpec)
```
This wasn't used at all.

nkolev92 committed:
cleanup, add package type to equals and hash code

nkolev92 committed:
kill processes in the test directory

nkolev92 committed:
cleanup

nkolev92 committed:
add a log message for compat check skipping of ridless graphs

nkolev92 committed:
Fix the package types reading and nuspec creation for test package

nkolev92 committed:
add an option to specify the probing path, tests

nkolev92 committed:
update the compat checker to allow for more than 1 package

nkolev92 committed:
add tests for the platforms package

nkolev92 committed:
fix isNullOrEmpty condition

nkolev92 committed:
add more tests, rework compat checker

nkolev92 committed:
fix the error message tests

nkolev92 committed:
Add a test for a package with multiple package types

nkolev92 committed:
localize all error messages

nkolev92 committed:
fix up rebase

nkolev92 committed:
remove line

emgarten commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -573,8 +574,8 @@ Copyright (c) .NET Foundation. All rights reserved.
       <_RestoreProjectName>$(MSBuildProjectName)</_RestoreProjectName>
-      <_RestoreProjectName Condition=" '$(RestoreProjectStyle)' == 'PackageReference' AND '$(AssemblyName)' != '' ">$(AssemblyName)</_RestoreProjectName>
-      <_RestoreProjectName Condition=" '$(RestoreProjectStyle)' == 'PackageReference' AND '$(PackageId)' != '' ">$(PackageId)</_RestoreProjectName>
+      <_RestoreProjectName Condition=" ('$(RestoreProjectStyle)' == 'PackageReference' OR '$(RestoreProjectStyle)' == 'DotnetToolReference') AND '$(AssemblyName)' != '' ">$(AssemblyName)</_RestoreProjectName>
```
This is used a lot, I suggest storing the value of `('$(RestoreProjectStyle)' == 'PackageReference' OR '$(RestoreProjectStyle)' == 'DotnetToolReference')` as a property and re-using it for these conditions. It can be done local to this target.

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+            contentItems.Load(compatibilityData.Files);
+
+
+            if (compatibilityData.TargetLibrary.PackageType.Any(e => e.Equals(PackageType.DotnetTool)))
```
`.Contains(PackageType.DotnetTool)` might be a better fit than `Any`

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+                    new PackageIdentity(node.Key.Name, node.Key.Version));
+
+                issues.Add(issue);
+                await _log.LogAsync(GetErrorMessage(NuGetLogCode.NU1204, issue, graph));//
```
remove comment

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+            }
+
+            if (containsDotnetToolPackageType &&
+                    !(HasCompatibleAssets(compatibilityData.TargetLibrary) || !compatibilityData.Files.Any(p => p.StartsWith("tools/", StringComparison.OrdinalIgnoreCase))))
```
Why allow non-tools compatible assets? Shouldn't this only look at the tools folder for a package of that type?

emgarten replied:
I'm not sure I understand this method, should there be one path for tool projects and another for everything else?

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+
+            if (ProjectStyle.DotnetToolReference == compatibilityData.PackageSpec.RestoreMetadata?.ProjectStyle)
+            {
+                if (compatibilityData.PackageSpec.GetAllPackageDependencies().Any(e => e.Name.Equals(compatibilityData.TargetLibrary.Name)))
```
Use ignore case here

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -324,10 +417,11 @@ internal static bool HasCompatibleAssets(LockFileTargetLibrary targetLibrary)
                 targetLibrary.Build.Count > 0 ||                                      // Build
-                targetLibrary.BuildMultiTargeting.Count > 0;                          // Cross targeting build
+                targetLibrary.BuildMultiTargeting.Count > 0 ||                        // Cross targeting build
+                targetLibrary.ToolsAssemblies.Count > 0;                              // Tools assemblies - This makes a package not backwards compatible potentially, but do tools assets matter
```
This should be separate check, a regular project should not see tool packages as compatible. I am guessing that this wouldn't be populated if it wasn't a tools project, but it would be more clear if there were a different method to check tool assets.

nkolev92 commented:
🔔 @emgarten 

// cc
@jainaashish @rohit21agrawal @mishra14

emgarten commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -554,12 +555,15 @@ Copyright (c) .NET Foundation. All rights reserved.
+    </PropertyGroup>
     <!-- Determine the restore output path -->
-    <PropertyGroup Condition=" '$(RestoreProjectStyle)' == 'PackageReference' OR '$(RestoreProjectStyle)' == 'ProjectJson' ">
+    <PropertyGroup Condition=" '$(PackageReferenceCompatibleProjectStyle)' == 'true' OR '$(RestoreProjectStyle)' == 'ProjectJson' ">
```
👍

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/ProjectStyle.cs:
```diff
@@ -21,18 +21,23 @@ public enum ProjectStyle : ushort
+        /// DotnetToolReference project
         /// </summary>
-        DotnetCliTool = 3,
+        DotnetToolReference = 3,
```
It shouldn't matter since I don't think anyone is using this enum where there would be a mix of NuGet.*.dll versions, but I'd typically expect this to be added at the end and not change the numbers of existing values.

emgarten commented on src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs:
```diff
@@ -502,6 +507,17 @@ internal ManagedCodePatterns(ManagedCodeConventions conventions)
+                        },
+                    pathPatterns: new PatternDefinition[]
+                        {
+                            new PatternDefinition("tools/{tfm}/{rid}/{any}", table: DotnetAnyTable),
```
Use `DefaultTfmAny` instead of `DotnetAnyTable` for both. At this point `dotnet` is deprecated so it would cause problems to use that for netstandard. 

You also want the real `Any` like `build/` uses.
 
You'll need to update the tests around this also.

emgarten commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetToolTests.cs:
```diff
@@ -0,0 +1,482 @@
+        [PlatformTheory(Platform.Windows)]
+        [InlineData("net461")]
+        [InlineData("netcoreapp1.0")]
+        public void DotnetToolTests_NoPackageReferenceToolRestore_ThrowsError(string tfm)
```
🏆 great tests

emgarten commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuilldIntegrationTestFixture.cs:
```diff
@@ -13,6 +13,7 @@
 using NuGet.Packaging.PackageExtraction;
 using NuGet.Protocol;
 using Xunit;
+using NuGet.Packaging.Core;
```
sort

jainaashish commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -468,6 +468,7 @@ Copyright (c) .NET Foundation. All rights reserved.
      RestorePackagesPath="$(RestorePackagesPath)"
      RestoreFallbackFolders="$(RestoreFallbackFolders)"
      RestoreConfigFile="$(RestoreConfigFile)"
+     RestoreRootConfigDirectory="$(RestoreRootConfigDirectory)"
```
do we wanna support this `RestoreRootConfigDirectory` via commandLine restore as well?

nkolev92 replied:
What do you mean by commandline?
People can set the MSBuild property if they want to, because we can't really hide it. 
It's not called in the normal nuget.exe/dotnet.exe workflows.

jainaashish commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:
```diff
@@ -470,32 +471,40 @@ private static Task LogDowngradeWarningsAsync(IEnumerable<RestoreTargetGraph> gr
                 {
-                    await logger.LogAsync(LogLevel.Verbose, string.Format(CultureInfo.CurrentCulture, Strings.Log_CheckingCompatibility, graph.Name));
+                    if (!(ProjectStyle.DotnetToolReference == project.RestoreMetadata?.ProjectStyle && string.IsNullOrEmpty(graph.RuntimeIdentifier)))
+                    { // Don't do compat checks for the ridless graph of DotnetTooReference restore. Everything relevant will be caught in the graph with the rid
```
nit: next line

jainaashish reviewed (approved):
Looks good to me.

nkolev92 commented:
@emgarten 
🔔 

I'll do the change for nested folders separately. 

Don't want to overextend this PR.

jainaashish commented:
Just out of curiosity, how this handles local tool installation?

nkolev92 commented:
@jainaashish 

From a NuGet perspective, there is no design/implementation work on global vs local (not even sure what/how they're doing that, it's still a WIP). 

There's 2 difference between global and local tools...how the tools can be invoked and where they are installed. 

The global tools, are as the name suggests, global, can be invoked from anywhere, so the tool location, which is at a global location, is added to the global path.
Each global tool will be installed in a self contained location, which is controlled by what the global packages folder for the restore will be. 

The local tools are just meant to be invokable in a certain context, and they'll be installed to the global packages folder of the user.

So the only thing that changes for us, is the global packages folder, which is a core scenario of ours anyways.

nkolev92 committed:
any framework rather than dotnet for tools, separate compatible assets check for tools vs dependency packages

rohit21agrawal commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -508,7 +509,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   -->
   <Target Name="_GetRestoreSettingsAllFrameworks"
-    Condition=" '$(TargetFrameworks)' != '' AND '$(RestoreProjectStyle)' == 'PackageReference' "
+    Condition=" '$(TargetFrameworks)' != '' AND ( '$(RestoreProjectStyle)' == 'PackageReference' OR '$(RestoreProjectStyle)' == 'DotnetToolReference' ) "
```
why not use PackageReferenceCompatibleProjectStyle here?

rohit21agrawal commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -496,7 +497,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   -->
   <Target Name="_GetRestoreSettingsCurrentProject"
-    Condition=" '$(TargetFrameworks)' == '' AND '$(RestoreProjectStyle)' == 'PackageReference' "
+    Condition=" '$(TargetFrameworks)' == '' AND ( '$(RestoreProjectStyle)' == 'PackageReference' OR '$(RestoreProjectStyle)' == 'DotnetToolReference' ) "
```
PackageReferenceCompatibleProjectStyle

nkolev92 replied:
The scope of that property is local.

rohit21agrawal replied:
as long as the target in which it is defined is executed before this one, that property can be used. it's much easier to maintain if there is only one place to add a future PackageReference compatible project style

nkolev92 replied:
Refactored it. 

I was more comfortable to just add it to the _GetRestoreProjectStyle returns list.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -60,22 +61,33 @@ public CompatibilityChecker(IReadOnlyList<NuGetv3LocalRepository> localRepositor
 
+            if (packageSpec.RestoreMetadata?.ProjectStyle == ProjectStyle.DotnetToolReference)
+            {
+                if (packageSpec.GetAllPackageDependencies().Count() != 1)
```
can this return a null list for zero package dependencies?

nkolev92 replied:
No, it'll return an empty set.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -236,7 +251,7 @@ private static RestoreLogMessage GetErrorMessage(NuGetLogCode logCode, Compatibi
                 graph.Conventions.Patterns.CompileRefAssemblies,
                 graph.Conventions.Patterns.RuntimeAssemblies,
-                graph.Conventions.Patterns.ContentFiles
+                graph.Conventions.Patterns.ContentFiles,
```
why the , in the end?

nkolev92 replied:
Stuff was added, then removed in a later commit, but the , got left over.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+                    group.Properties.TryGetValue(ManagedCodeConventions.PropertyNames.TargetFrameworkMoniker, out var tfmObj);
+
+                    var tfm = tfmObj as NuGetFramework;
+                    var rid = ridObj as string;
```
you are assumbing TryGetValue returned the right value for both of these, is that a safe assumption?

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+                    group.Properties.TryGetValue(ManagedCodeConventions.PropertyNames.RuntimeIdentifier, out var ridObj);
+                    group.Properties.TryGetValue(ManagedCodeConventions.PropertyNames.TargetFrameworkMoniker, out var tfmObj);
+
+                    var tfm = tfmObj as NuGetFramework;
```
you are assumbing TryGetValue returned the right value for both of these, is that a safe assumption?

nkolev92 replied:
What do you mean by the right value?
Nulls are ok, won't cause any problems, but that'll be caught way earlier.
This is mostly for tests. 
We already do the same thing on L269

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+            }
+
+            if (containsDotnetToolPackageType &&
+                    !(HasCompatibleToolsAssets(compatibilityData.TargetLibrary) || !compatibilityData.Files.Any(p => p.StartsWith("tools/", StringComparison.OrdinalIgnoreCase))))
```
since you do a check for files fairly often, can we abstract out functions that would tell us if there are any lib files or tool files like this: 

compatibilityData.HasLibFiles
compatibilityData.HasToolsFiles etc etc.

nkolev92 replied:
HasCompatibleToolsAssetts is a static used once in the compat checker, and once in the lock file builder. 
It can't be bound to the CompatibilityData model.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
                 !compatibilityData.Files.Any(p =>
                     p.StartsWith("ref/", StringComparison.OrdinalIgnoreCase)
-                    || p.StartsWith("lib/", StringComparison.OrdinalIgnoreCase));                       // No assemblies at all (for any TxM)
+                    || p.StartsWith("lib/", StringComparison.OrdinalIgnoreCase)); // No assemblies at all (for any TxM)
```
see : https://github.com/NuGet/NuGet.Client/pull/1900/files#r161083060

nkolev92 replied:
I'd rather not refactor code that doesn't need changing. 
Same as above, can't tie to the model.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityChecker.cs:
```diff
@@ -301,16 +316,94 @@ private static bool IsProjectCompatible(Library library)
+            {
+                if (compatibilityData.PackageSpec.GetAllPackageDependencies().Any(e => e.Name.Equals(compatibilityData.TargetLibrary.Name, StringComparison.OrdinalIgnoreCase)))
+                {
+                    if (!containsDotnetToolPackageType)
```
change this to :
if(!containsDotnetToolPackageType && compatibilityData.PackageSpec.GetAllPackageDependencies().Any(e => e.Name.Equals(compatibilityData.TargetLibrary.Name, StringComparison.OrdinalIgnoreCase)))

this way it will not have to evaluate the more expensive check if containsDotnetToolPackageType is false.

nkolev92 replied:
Good catch. Not sure how I missed that simplification.

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityIssue.cs:
```diff
@@ -59,7 +63,44 @@ public static CompatibilityIssue ReferenceAssemblyNotImplemented(string assembly
+                CompatibilityIssueType.IncompatiblePackageWithDotnetTool,
+                referenceAssemblyPackage,
+                string.Empty,
+                null,
```
can you name the null parameters for readability?

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityIssue.cs:
```diff
@@ -59,7 +63,44 @@ public static CompatibilityIssue ReferenceAssemblyNotImplemented(string assembly
+                CompatibilityIssueType.ToolsPackageWithExtraPackageTypes,
+                referenceAssemblyPackage,
+                string.Empty,
+                null,
```
name the null parameters

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityIssue.cs:
```diff
@@ -74,7 +115,21 @@ public static CompatibilityIssue ReferenceAssemblyNotImplemented(string assembly
+                CompatibilityIssueType.ProjectWithIncorrectDependencyCount,
+                project,
+                string.Empty,
+                null,
```
name null parameters

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs:
```diff
@@ -49,9 +49,10 @@ public class LockFileBuilder
             var previousLibraries = previousLockFile?.Libraries.ToDictionary(l => Tuple.Create(l.Name, l.Version));
 
-            if (project.RestoreMetadata?.ProjectStyle == ProjectStyle.PackageReference)
+            if (project.RestoreMetadata?.ProjectStyle == ProjectStyle.PackageReference ||
```
suggestion: have a property wrapper like IsPackageReferenceCompatibleProjectStyle

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs:
```diff
@@ -169,6 +169,7 @@ public void ApplyStandardProperties(RestoreRequest request)
             request.PackageSaveMode = PackageSaveMode;
 
             if (request.ProjectStyle == ProjectStyle.PackageReference
+                || request.ProjectStyle == ProjectStyle.DotnetToolReference
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:
```diff
@@ -253,6 +253,7 @@ private string GetAssetsFilePath(LockFile lockFile)
             if (string.IsNullOrEmpty(projectLockFilePath))
             {
                 if (_request.ProjectStyle == ProjectStyle.PackageReference
+                    || _request.ProjectStyle == ProjectStyle.DotnetToolReference
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs:
```diff
@@ -470,32 +471,40 @@ private static Task LogDowngradeWarningsAsync(IEnumerable<RestoreTargetGraph> gr
+                                await logger.LogAsync(LogLevel.Debug, $"Incompatible packages: {packageCount}");
+                            }
                         }
+                    } else
```
nit: format doc

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs:
```diff
@@ -365,7 +365,7 @@ public static string GetMSBuildFilePath(PackageSpec project, RestoreRequest requ
             string path;
 
-            if (request.ProjectStyle == ProjectStyle.PackageReference)
+            if (request.ProjectStyle == ProjectStyle.PackageReference || request.ProjectStyle == ProjectStyle.DotnetToolReference) 
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -93,7 +93,8 @@ public static DependencyGraphSpec GetDependencySpec(IEnumerable<IMSBuildItem> it
                     || spec.RestoreMetadata.ProjectStyle == ProjectStyle.ProjectJson
                     || spec.RestoreMetadata.ProjectStyle == ProjectStyle.DotnetCliTool
-                    || spec.RestoreMetadata.ProjectStyle == ProjectStyle.Standalone)
+                    || spec.RestoreMetadata.ProjectStyle == ProjectStyle.Standalone
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -194,7 +195,8 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
                     || restoreType == ProjectStyle.Standalone
                     || restoreType == ProjectStyle.DotnetCliTool
-                    || restoreType == ProjectStyle.ProjectJson)
+                    || restoreType == ProjectStyle.ProjectJson
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -222,7 +224,8 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
                 if (restoreType == ProjectStyle.PackageReference
                     || restoreType == ProjectStyle.Standalone
-                    || restoreType == ProjectStyle.DotnetCliTool)
+                    || restoreType == ProjectStyle.DotnetCliTool
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -235,7 +238,8 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
 
                 if (restoreType == ProjectStyle.PackageReference
-                    || restoreType == ProjectStyle.Standalone)
+                    || restoreType == ProjectStyle.Standalone
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:
```diff
@@ -274,7 +278,8 @@ public static PackageSpec GetPackageSpec(IEnumerable<IMSBuildItem> items)
                     || restoreType == ProjectStyle.Unknown
-                    || restoreType == ProjectStyle.PackagesConfig)
+                    || restoreType == ProjectStyle.PackagesConfig
+                    || restoreType == ProjectStyle.DotnetToolReference)
```
https://github.com/NuGet/NuGet.Client/pull/1900/files#r161085303

rohit21agrawal commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuilldIntegrationTestFixture.cs:
```diff
@@ -60,32 +61,86 @@ internal void CreateDotnetNewProject(string solutionRoot, string projectName, st
 
-            Assert.True(result.Item1 == 0, $"Creating project failed with following log information :\n {result.AllOutput}");
-            Assert.True(string.IsNullOrWhiteSpace(result.Item3), $"Creating project failed with following message in error stream :\n {result.AllOutput}");
+            var projectFile = $@"<Project Sdk=""Microsoft.NET.Sdk"">
```
is this the only way to create a dotnet tool project? it's ugly :(

nkolev92 replied:
Discussed offline.

rohit21agrawal commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuilldIntegrationTestFixture.cs:
```diff
@@ -60,32 +61,86 @@ internal void CreateDotnetNewProject(string solutionRoot, string projectName, st
+            Assert.True(File.Exists(projectFileName));
+        }
+
+        internal CommandRunnerResult RestoreToolProject(string workingDirectory, string projectName, string args = "")
```
why is this separate from RestoreProject?

rohit21agrawal commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuilldIntegrationTestFixture.cs:
```diff
@@ -209,6 +335,11 @@ private static void KillDotnetExe(string pathToDotnetExe)
                         }
                     }
+
+                    foreach (var process in testDirProcesses)
```
remove this

nkolev92 replied:
Discussed offline.

rohit21agrawal reviewed (changes requested):
mostly looks great, some feedback , questions and suggestions left in comments

nkolev92 committed:
address feedback

nkolev92 committed:
refactor out the targets to do less of the if package reference and dotnet tool reference

nkolev92 committed:
don't run the tests for recursive

nkolev92 committed:
cleanup

nkolev92 commented:
🔔 
@rohit21agrawal - Addressed feedback. I will wait for your merge for the MsBuildIntegrationFixture cleanup and then rebase before I push. 
@emgarten

rohit21agrawal commented on src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:
```diff
@@ -457,7 +460,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   -->
   <Target Name="_GetRestoreSettings"
-          Condition=" '$(RestoreProjectStyle)' == 'PackageReference' OR '$(RestoreProjectStyle)' == 'ProjectJson' "
+          Condition=" '$(RestoreProjectStyle)' == 'PackageReference' OR '$(RestoreProjectStyle)' == 'ProjectJson' OR '$(RestoreProjectStyle)' == 'DotnetToolReference' "
```
missed PackageReferenceCompatibleStyle here

emgarten commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/CompatibilityIssue.cs:
```diff
@@ -168,6 +271,27 @@ private string FormatMessage(string message, string supports, string noSupports)
+                        .OrderBy(s => s, StringComparer.CurrentCultureIgnoreCase))
+                    {
+                        sb.Append(Environment.NewLine);
+                        sb.Append($"  - {framework}");
```
nit: there might be a helper around for building these multi-line message strings

emgarten commented on src/NuGet.Core/NuGet.Packaging/ContentModel/ManagedCodeConventions.cs:
```diff
@@ -55,6 +55,14 @@ public class ManagedCodeConventions
+            new PatternTableEntry(
+                PropertyNames.TargetFrameworkMoniker,
+                "any",
+                AnyFramework.Instance )
```
💯

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileTargetLibrary.cs:
```diff
@@ -65,7 +70,9 @@ public bool Equals(LockFileTargetLibrary other)
-                && BuildMultiTargeting.OrderedEquals(other.BuildMultiTargeting, item => item.Path, StringComparer.OrdinalIgnoreCase);
+                && BuildMultiTargeting.OrderedEquals(other.BuildMultiTargeting, item => item.Path, StringComparer.OrdinalIgnoreCase)
+                && ToolsAssemblies.OrderedEquals(other.ToolsAssemblies, item => item.Path, StringComparer.OrdinalIgnoreCase)
+                && PackageType.OrderedEquals(other.PackageType, item => item);
```
Does this cause problems with round tripping? For example if the in-memory lock file has package types set but the on disk one did not write this, does this still work or does it end up re-writing the file?

nkolev92 replied:
Great call. 

It does. 
I'll look into adding some tests to catch these scenarios.

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileTargetLibrary.cs:
```diff
@@ -132,6 +139,16 @@ public override int GetHashCode()
                 combiner.AddObject(item);
             }
 
+            foreach (var item in ToolsAssemblies.OrderBy(item => item.Path, StringComparer.OrdinalIgnoreCase))
```
If case does matter then then use Ordinal sort, if not I would expect it to hash with ignore case

nkolev92 replied:
The "item" element is LockFileItem....the lockFileItem hashcode does  the following:
```
        public override int GetHashCode()
        {
            var combiner = new HashCodeCombiner();

            combiner.AddStringIgnoreCase(Path);
            combiner.AddDictionary(Properties);

            return combiner.CombinedHash;
        }
```

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileTargetLibrary.cs:
```diff
@@ -132,6 +139,16 @@ public override int GetHashCode()
+                combiner.AddObject(item);
+            }
+
+            foreach(var item in PackageType)
```
sorting for this?

nkolev92 replied:
Gone now.

nkolev92 committed:
address feedback

nkolev92 committed:
Don't add package type to the equals and hash code checks

nkolev92 committed:
fix compile error

nkolev92 committed:
Add file nesting

nkolev92 committed:
fix tests, add more nested values tests

nkolev92 commented:
🔔 @emgarten 

Addressed the feedback. Also added support for nested files in the tools directory.

nkolev92 committed:
refactor the msbuild integration test fixture

emgarten commented on src/NuGet.Core/NuGet.Commands/Strings.resx:
```diff
@@ -682,4 +682,16 @@ For more information, visit http://docs.nuget.org/docs/reference/command-line-re
     <comment>{0} is the certificate store name</comment>
   </data>
+  <data name="Error_ProjectWithIncorrectDependenciesCount" xml:space="preserve">
+    <value>Project {0} must have exactly {1} package reference(s).</value>
```
nit: this could just say a single package reference, I'm not sure that writing in the number is needed.

emgarten commented on src/NuGet.Core/NuGet.Commands/Strings.resx:
```diff
@@ -682,4 +682,16 @@ For more information, visit http://docs.nuget.org/docs/reference/command-line-re
+    <value>Invalid project-package combination for {0} {1}. DotnetToolReference project style can only contain references of the DotnetTool type</value>
+  </data>
+  <data name="Error_ToolsPackageWithExtraPackageTypes" xml:space="preserve">
+    <value>Invalid tools package {0} {1}. Tools packages cannot contain more than 1 PackageType.</value>
```
nit: as a general rule numbers smaller than 10 should be written out as the word, not as the numeral. So: *more than one PackageType* here

emgarten commented on src/NuGet.Core/NuGet.Packaging.Core/PackageType.cs:
```diff
@@ -87,5 +89,28 @@ public override int GetHashCode()
+
+            var res = StringComparer.OrdinalIgnoreCase.Compare(Name, other.Name);
+
+            if(res != 0)
```
nit: spacing

emgarten commented on src/NuGet.Core/NuGet.Packaging.Core/PackageType.cs:
```diff
@@ -87,5 +89,28 @@ public override int GetHashCode()
+                return res;
+            }
+
+            return Version.CompareTo(other.Version);
```
Looks like Version could be null here and throw. Maybe the parser prevents this, but this class seems to allow it.

nkolev92 replied:
There's a null version check, so version can not be null.

emgarten commented on test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuilldIntegrationTestFixture.cs:
```diff
@@ -169,6 +218,75 @@ private void UpdateCliWithLatestNuGetAssemblies(string cliDirectory)
+                }
+                try
+                { 
+                File.Delete(targetFilePath);
```
nit: format the file

emgarten commented on test/NuGet.Core.Tests/NuGet.Packaging.Test/ContentModelTests/ContentModelToolsTests.cs:
```diff
@@ -0,0 +1,218 @@
+
+            // Assert
+            Assert.Equal(1, groups.Count);
+            Assert.Equal(NuGetFramework.AnyFramework, (NuGetFramework)groups.First().Properties["tfm"]);
```
🥇 good test

emgarten reviewed (approved):
Minor comments. This looks very complete and with lots of tests great! 👏

nkolev92 committed:
Address feedback
